cert-tool-src/ech.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/crypto/cryptobyte"
+
+	"github.com/cloudflare/circl/hpke"
+)
+
+const (
+	ECHVersionDraft09 uint16 = 0xff09 // draft-ietf-tls-esni-09
+)
+
+// makeECHKey generates an ECH config and corresponding key, writing the key to
+// outKeyPath and the config to outPath. The config is encoded as an ECHConfigs
+// structure as specified by draft-ietf-tls-esni-09 (i.e., it is prefixed by
+// 16-bit integer that encodes its length). This is the format as it is consumed
+// by the client.
+//
+// The first DNS name of certificate inCertPath is used as the ECH config's
+// public name. The remaining parameters of the ECH config are specified by the
+// template.
+func makeECHKey(template ECHConfigTemplate, inCertPath, outPath, outKeyPath string) {
+	certParseErrMsg := "failed to parse input certificate"
+	keyGenErrMsg := "failed to generate ECH key"
+
+	// Load the certificate of the client-facing server.
+	clientFacingCertPEMBlock, err := ioutil.ReadFile(filepath.Join(inCertPath))
+	fatalIfErr(err, certParseErrMsg)
+	clientFacingCertDERBlock, _ := pem.Decode(clientFacingCertPEMBlock)
+	if clientFacingCertDERBlock == nil || clientFacingCertDERBlock.Type != "CERTIFICATE" {
+		fatalIfErr(errors.New("PEM decoding failed"), certParseErrMsg)
+	}
+	clientFacingCert, err := x509.ParseCertificate(clientFacingCertDERBlock.Bytes)
+	fatalIfErr(err, certParseErrMsg)
+
+	// Generate the ECH config and corresponding key, using the first DNS name
+	// specified by the client-facing certificate as the public name.
+	if len(clientFacingCert.DNSNames) == 0 {
+		fatalIfErr(errors.New("input certificate does not specify a DNS name"), keyGenErrMsg)
+	}
+	template.PublicName = clientFacingCert.DNSNames[0]
```

Comment:
This might not be a valid public facing name (*.example.com) -- should we check this rather than just picking the first one?

Reply:
Hrm ... good point. I will fix this.

Reply:
Well, `DNSNames[0]` is hard-coded as "server" here, which is already not a valid public name. (It needs to be an SNI.) I just decided to add this as a TODO and will address it in #20.

Reply:
That works!